dashboard/app/models/script.rb

```diff
@@ -437,14 +437,18 @@ def self.get_from_cache(id_or_name, version_year: nil)
   end
 
   # Given a script family name, return a dummy Script with redirect_to field
-  # pointing toward the latest stable script in that family, or to a specific
-  # version_year if one is specified.
+  # pointing toward the latest stable script assigned to the user in that family, or to a specific
+  # version_year and locale if specified.
   # @param family_name [String] The name of the script family to search in.
-  # @param version_year [String] Version year to return. Optional.
+  # @param version_year [String] Optional. Version year to return.
+  # @param user [User] Optional. If provided, will attempt to retrieve latest stable script
+  # that the user is assigned to or has progress in.
+  # @param locale [String] Optional. Locale of the request.
   # @return [Script|nil] A dummy script object, not persisted to the database,
   #   with only the redirect_to field set.
-  def self.get_script_family_redirect(family_name, version_year: nil)
-    script_name = Script.latest_stable_version(family_name, version_year: version_year).try(:name)
+  def self.get_script_family_redirect(family_name, version_year: nil, user: nil, locale: 'en-US')
+    script_name = Script.latest_assigned_version(family_name, user).try(:name)
+    script_name ||= Script.latest_stable_version(family_name, version_year: version_year, locale: locale, fallback: true).try(:name)
```

Comment:
I'm afraid we can't afford to do this either -- this lets each call to `ScriptsController#show` hit our database. We need this method to rely on the script cache. 

A problem you may run into is that `get_from_cache` calls `get_without_cache` which will `raise` if the script is not found, which will happen when you are looking for a family name. Here is what I would recommend:
* Create a whitelist of script family names in script_constants.rb. When asked to `show` one of these, look for all scripts with this family name rather than scripts with this name, and use locale and user to decide which to redirect to.
* when looking for scripts by family name, you will need to cache the result so that we don't hit the database every time someone goes to /s/coursea (for example). I would imagine that you'd define this `script_family_from_cache` and `script_family_without_cache`, which takes in a family name and returns an array of scripts belonging to that family.
* since this will be a new kind of cache entry, there will be some places in our setup code that these need to be populated. 
* we have some unit tests which count how many database requests we are making on certain controller routes once the script cache has been populated. those are great to run as you iterate to make sure that you aren't adding too many DB requests, so that we don't have to reason abstractly about cache hits and misses.

Reply:
i just realized i never thanked you for this guidance 😄 thank you! this was really helpful